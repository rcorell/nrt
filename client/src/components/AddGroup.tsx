import { useMutation } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { createGroupMutationString } from 'src/api/api';
import { CreateGroupMutation, CreateGroupMutationVariables } from 'src/api/__generated__/CreateGroupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    setBrowserTitle('Add Group');

    const [createGroup, { error }] = useMutation<CreateGroupMutation, CreateGroupMutationVariables>(
        createGroupMutationString,
        {
            variables: { description, name },
            onCompleted: ({ createGroup }) => {
                createGroup.id += 0;
            }
        }
    );

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createGroup();
    };

    const errorDisplay = () => {
        if (error) {
            return <div>Error: {JSON.stringify(error)}</div>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label>Name</Form.Label>
                    <Form.Control
                        autoFocus
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Description</Form.Label>
                    <Form.Control
                        id="description"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
