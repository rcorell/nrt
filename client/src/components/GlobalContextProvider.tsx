import React, { useState } from 'react';

import { isAuthenticated } from 'src/api/api';

type GlobalContextType = {
    authenticated: boolean;
    setAuthenticated: (authenticated: boolean) => void;
};

export const GlobalContext = React.createContext<GlobalContextType>({
    authenticated: false,
    setAuthenticated: (_authenticated) => {}
});

export const GlobalContextProvider: React.FC = ({ children }) => {
    const [authenticated, setAuthenticated] = useState(isAuthenticated());

    const contextFunctions = { authenticated, setAuthenticated };

    return <GlobalContext.Provider value={contextFunctions}>{children}</GlobalContext.Provider>;
};
