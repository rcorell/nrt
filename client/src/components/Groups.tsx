import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
