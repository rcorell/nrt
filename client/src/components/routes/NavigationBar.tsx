import * as React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import * as loginActions from '../../redux/login/actions';
import { getIsAuthenticated } from '../../redux/login/selectors';
import * as navigationActions from '../../redux/navigation/actions';
import { Path } from '../../redux/navigation/types';
import { State } from '../../redux/state.types';
import { NavBar, NavBarLeft, NavBarRight } from './styles';

interface NavigationBarStateProps {
    authenticated: boolean;
}

interface NavigationBarDispatchProps {
    addTopic: () => void;
    logout: () => void;
    home: () => void;
    signup: () => void;
}

export type NavigationBarComponentProps = NavigationBarStateProps & NavigationBarDispatchProps;

export class NavigationBarComponent extends React.Component<NavigationBarComponentProps> {
    render() {
        let leftNavBar;
        let rightNavBar;

        if (this.props.authenticated) {
            leftNavBar = (
                <>
                    <Link to="" onClick={() => this.props.home()}>
                        Home
                    </Link>
                    &nbsp;|&nbsp;
                    <Link to="" onClick={() => this.props.addTopic()}>
                        Add Topic
                    </Link>
                </>
            );
            rightNavBar = (
                <Link to="" onClick={() => this.props.logout()}>
                    Logout
                </Link>
            );
        } else {
            leftNavBar = (
                <>
                    <Link to={Path.LOGIN}>Log in</Link>
                    &nbsp;|&nbsp;
                    <Link to={Path.SIGNUP}>Sign up</Link>
                </>
            );
        }

        return (
            <>
                <NavBar>
                    <NavBarLeft>{leftNavBar}</NavBarLeft>
                    <NavBarRight>{rightNavBar}</NavBarRight>
                </NavBar>
            </>
        );
    }
}

export const mapStateToProps: (state: State) => NavigationBarStateProps = (state: State) => ({
    authenticated: getIsAuthenticated(state)
});

export const mapDispatchToProps = (dispatch: Function) => ({
    addTopic: () => dispatch(navigationActions.addTopic()),
    logout: () => dispatch(loginActions.logout()),
    home: () => dispatch(navigationActions.home()),
    signup: () => dispatch(navigationActions.signup())
});

export const NavigationBar = connect(mapStateToProps, mapDispatchToProps)(NavigationBarComponent);
