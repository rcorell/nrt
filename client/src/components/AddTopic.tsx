import { useMutation, useQuery } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { CreateTopicMutation, CreateTopicMutationVariables } from 'src/api/__generated__/CreateTopicMutation';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { createTopicMutationString, fetchUserQueryString } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const AddTopic: React.FC = () => {
    const [description, setDescription] = useState('');
    const [groupId, setGroupId] = useState('');
    const [title, setTitle] = useState('');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);

    const [createTopic, { error }] = useMutation<CreateTopicMutation, CreateTopicMutationVariables>(
        createTopicMutationString,
        {
            variables: { groupId, title, description },
            onError: () => {
                // RTL bug
            }
        }
    );

    const isFormInvalid = () => {
        return title.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createTopic();
    };

    const errorStatus = () => {
        if (error) {
            return <div>Error: ${JSON.stringify(error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Topic');

    if (userQuery.loading || !userQuery.data) {
        return <div>{LOADING_TEXT}</div>;
    }

    const groupOptions = userQuery.data.user.groups.map((group) => (
        <option key={group!.id} value={group!.id}>
            {group!.name}
        </option>
    ));

    return (
        <FormContainer>
            <h1>Add Topic</h1>
            {errorStatus()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="groupLabel">Group</Form.Label>
                    <Form.Control
                        as="select"
                        aria-labelledby="groupLabel"
                        id="group"
                        onChange={(e) => setGroupId(e.target.value)}
                        size="lg"
                    >
                        {groupOptions}
                    </Form.Control>
                </Form.Group>

                <Form.Group>
                    <Form.Label id="topicLabel">Topic</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="topicLabel"
                        id="topic"
                        onChange={(e) => setTitle(e.target.value)}
                        size="lg"
                        value={title}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Submit Topic
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
