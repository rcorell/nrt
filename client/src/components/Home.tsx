import { useQuery } from '@apollo/client';
import React from 'react';

import {
    FetchUserWithGroupTopicsQuery,
    FetchUserWithGroupTopicsQuery_user_groups
} from 'src/api/__generated__/FetchUserWithGroupTopicsQuery';
import { fetchUserWithGroupTopicsQuery } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { HomePage } from 'src/styles/home';
import { setBrowserTitle } from 'src/utils';

export const Home: React.FC = () => {
    const query = useQuery<FetchUserWithGroupTopicsQuery>(fetchUserWithGroupTopicsQuery);

    setBrowserTitle('Home');

    if (query.loading) {
        return <b>{LOADING_TEXT}</b>;
    }
    if (query.error) {
        return <div>{JSON.stringify(query.error)}</div>;
    }

    const groupTopics = (group: FetchUserWithGroupTopicsQuery_user_groups | null) => {
        return (
            <div key={group!.id}>
                <h2>{group!.name}</h2>
                <ul>
                    {group!.topics.map((topic) => (
                        <li key={topic!.id}>{topic!.title}</li>
                    ))}
                </ul>
            </div>
        );
    };

    return <HomePage>{query!.data!.user.groups.map((group) => groupTopics(group))}</HomePage>;
};
