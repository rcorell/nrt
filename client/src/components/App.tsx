import 'bootstrap/dist/css/bootstrap.min.css';

import { ApolloProvider } from '@apollo/client';
import React from 'react';

import { GlobalContextProvider } from 'src/components/GlobalContextProvider';
import { apolloClient } from 'src/api/api';
import { Routes } from 'src/components/Routes';
import { AppHeader, GlobalStyle } from 'src/styles';

export const App: React.FC = () => {
    return (
        <>
            <GlobalStyle />
            <GlobalContextProvider>
                <ApolloProvider client={apolloClient}>
                    <AppHeader>Top 5 Daily</AppHeader>
                    <Routes />
                </ApolloProvider>
            </GlobalContextProvider>
        </>
    );
};
