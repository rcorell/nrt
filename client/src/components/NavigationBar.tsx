import { A, navigate } from 'hookrouter';
import React, { useContext } from 'react';

import { logout } from 'src/api/api';
import { GlobalContext } from 'src/components/GlobalContextProvider';
import { Path } from 'src/components/Routes';
import { NavBar, NavBarLeft, NavBarRight } from 'src/styles/navbar';

export const NavigationBar: React.FC = () => {
    let leftNavBar;
    let rightNavBar;

    const { authenticated, setAuthenticated } = useContext(GlobalContext);

    const logoutAndGoToLoginPage = () => {
        logout();
        setAuthenticated(false);
        navigate(Path.LOGIN);
    };

    if (authenticated) {
        leftNavBar = (
            <>
                <A href={Path.HOME}>Home</A>
                &nbsp;|&nbsp;
                <A href={Path.TOPICS}>Topics</A>
                &nbsp;|&nbsp;
                <A href={Path.ADD_TOPIC}>Add Topic</A>
                &nbsp;|&nbsp;
                <A href={Path.GROUPS}>Groups</A>
                &nbsp;|&nbsp;
                <A href={Path.ADD_GROUP}>Add Group</A>
            </>
        );
        rightNavBar = (
            <A href="" onClick={logoutAndGoToLoginPage}>
                Logout
            </A>
        );
    } else {
        leftNavBar = (
            <>
                <A href={Path.LOGIN}>Log in</A>
                &nbsp;|&nbsp;
                <A href={Path.SIGNUP}>Sign up</A>
            </>
        );
        rightNavBar = null;
    }

    return (
        <NavBar>
            <NavBarLeft>{leftNavBar}</NavBarLeft>
            <NavBarRight>{rightNavBar}</NavBarRight>
        </NavBar>
    );
};
