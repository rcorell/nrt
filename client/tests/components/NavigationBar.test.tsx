import { NavigationBar } from 'src/components/NavigationBar';

describe('NavigationBar', () => {
    it('matches snapshot', () => {
        global.expectComponentToMatchSnapshot(NavigationBar);
    });
});
