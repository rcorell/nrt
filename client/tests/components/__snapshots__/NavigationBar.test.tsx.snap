// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar snapshots authenticated 1`] = `
<div>
  <div
    class="sc-iCfLBT kRgVyw"
  >
    <div
      class="sc-furvIG"
    >
      <a
        href="/"
      >
        Home
      </a>
       | 
      <a
        href="/topics"
      >
        Topics
      </a>
       | 
      <a
        href="/add-topic"
      >
        Add Topic
      </a>
       | 
      <a
        href="/groups"
      >
        Groups
      </a>
       | 
      <a
        href="/add-group"
      >
        Add Group
      </a>
    </div>
    <div
      class="sc-pVTma"
    >
      <a
        href="/user-profile"
      >
        Profile
      </a>
       | 
      <a
        href=""
      >
        Logout
      </a>
    </div>
  </div>
</div>
`;

exports[`NavigationBar snapshots unauthenticated 1`] = `
<div
  className="sc-iCfLBT kRgVyw"
>
  <div
    className="sc-furvIG"
  >
    <a
      href="/login"
      onClick={[Function]}
    >
      Log in
    </a>
     | 
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign up
    </a>
  </div>
  <div
    className="sc-pVTma"
  />
</div>
`;
